# Modulo 2: Autovectores y autovalores

# Autovectores y Autovalores

Los autovectores y autovalores son conceptos fundamentales en álgebra lineal y tienen aplicaciones importantes en diversas áreas, incluyendo la física, la ingeniería y la informática. Estas propiedades matemáticas se utilizan para entender y analizar transformaciones lineales, como la aplicación lineal T en este ejemplo.

###  Ahora en español fácil y sencilo para todas las edades: 

Imagina que tienes un mapa mágico, y este mapa puede mover objetos en el mundo. Un **autovector es como una flecha en el mapa que no cambia su dirección cuando se mueve**. En otras palabras, si tienes un autovector en el mapa y lo mueves, sigue apuntando en la misma dirección que antes.

Por ejemplo, si tienes una flecha que apunta hacia el norte en el mapa y lo mueves, seguirá apuntando hacia el norte. Esa flecha en el mapa es como un autovector.

Los autovectores son importantes en matemáticas y ciencia porque nos ayudan a entender cómo ciertas transformaciones o cambios afectan a los objetos. En el caso de los autovectores, nos dicen en qué dirección los objetos no cambian cuando se aplica una transformación.

Un **autovalor está relacionado con los autovectores y es como una especie de "magnitud" o "cantidad de cambio"** que se produce cuando aplicas una transformación a un autovector.

Siguiendo con la analogía del mapa mágico, un **autovalor sería como un número especial que te dice cuánto se estira o encoge una flecha (autovector)** cuando la transformación actúa sobre ella. Si el autovalor es grande, la flecha se estirará mucho; si es pequeño, se estirará solo un poco, o incluso se podría encoger.

### Nos volvemos a poner técnicos entonces: 

## Autovalores

Un autovalor de una matriz (o un operador lineal) es un número que caracteriza cómo se escala un autovector cuando se aplica la matriz o el operador. Es decir, un número λ es un autovalor de una matriz A si existe un vector no nulo v, llamado autovector, tal que la multiplicación de A por v resulta en un vector que es simplemente λ veces v.

Matemáticamente, esto se expresa de la siguiente manera:

$ A \cdot v = \lambda \cdot v $

Donde:
- \( A \) es la matriz (en nuestro caso, la matriz de la aplicación lineal T).
- \( v \) es el autovector.
- \( \lambda \) es el autovalor.

## Autovectores

Los autovectores son los vectores asociados a los autovalores y son fundamentales para comprender cómo las matrices o los operadores afectan la dirección y la magnitud de los vectores. Los autovectores son únicos hasta una constante de escala. Esto significa que, si \( v \) es un autovector, entonces cualquier múltiplo de \( v \) también es un autovector asociado al mismo autovalor \( \lambda \).

En el contexto de la aplicación lineal T(x, y) = (2x + y, -x + 3y), los autovectores nos ayudarán a entender cómo T transforma los vectores y, en particular, cómo se escalan.

En el siguiente apartado, vamos a calcular los autovalores y autovectores de la aplicación lineal T y verificar que cumplan con las propiedades mencionadas.

## Veamos un ejemplo:


In [12]:
#Vamso a considerar la matriz A
A = np.array([[3, 1],
              [1, 2]])

autovalores, autovectores = np.linalg.eig(A) #Esta función nos devuelve dos argumentos los autovalores y los autovectores

#Basicamente el método numpy.linalg.eig() lo hace todo por ti 

#Veamos que contiene la variable autovalores:

print(autovalores) #Hint: Es una lista con los autovalores 

[3.61803399 1.38196601]


Los números que obtuvimos como autovalores, [3.61803399, 1.38196601], tienen significados importantes en el contexto de la matriz A y sus aplicaciones lineales. Cada número representa un autovalor, y aquí está lo que significa:

* 3.61803399: Este es el primer autovalor. Indica cuánto se estira (o "amplía") el autovector correspondiente cuando se aplica la matriz A. En este caso, significa que el autovector correspondiente se ampliará aproximadamente 3.618 veces en la dirección dada por ese autovector.

* 1.38196601: Este es el segundo autovalor. Al igual que el primero, representa cuánto se estira (o "amplía") el segundo autovector cuando se aplica la matriz A. En este caso, el autovector correspondiente se ampliará aproximadamente 1.382 veces en su dirección.

Cada autovalor te dice cuánto cambia la magnitud de un autovector particular cuando se aplica la matriz. En este ejemplo, la matriz A estira los autovectores en direcciones específicas con los factores de escala dados por los autovalores.

In [11]:
print(autovectores)

[[ 0.85065081 -0.52573111]
 [ 0.52573111  0.85065081]]


Cada fila en la matriz de autovectores representa un autovector. En este caso, tenemos dos autovectores. Aquí hay una explicación de lo que representan:

* El primer autovector [0.85065081, -0.52573111] está asociado al primer autovalor 3.61803399. Esto significa que cuando aplicas la matriz, este autovector se estira (o se contrae) aproximadamente 3.618 veces en su dirección original.

* El segundo autovector [0.52573111, 0.85065081] está asociado al segundo autovalor 1.38196601. Esto significa que cuando aplicas la matriz, este autovector se estira (o se contrae) aproximadamente 1.381 veces en su dirección original.

Los autovectores nos indican las direcciones en las que los vectores no cambian de dirección, sino que pueden cambiar en magnitud (ampliarse o encogerse) cuando aplicamos la matriz. En este caso, estos autovectores específicos representan las direcciones en las que ocurre ese estiramiento o contracción.

In [9]:
print("Autovalores:", autovalores)
print("Autovectores:")
for i in range(len(autovectores)):
    print(f"Autovector {i + 1}:", autovectores[:, i])

Autovalores: [3.61803399 1.38196601]
Autovectores:
Autovector 1: [0.85065081 0.52573111]
Autovector 2: [-0.52573111  0.85065081]



Sea la aplicación lineal T(x,y)=(x',y')=(2x+y,-x+3y). Calcula los autovalores y autovectores de T con Python. Comprobar que realmente son autovalores y autovectores.

In [2]:
import numpy as np

In [3]:
# Definir la matriz de la aplicación lineal T
T_matrix = np.array([[2, 1],
                     [-1, 3]])

# Calcular los autovalores y autovectores
autovalores, autovectores = np.linalg.eig(T_matrix)

# Mostrar los resultados
print("Autovalores:", autovalores)
print("Autovectores:")
for i in range(len(autovectores)):
    print(f"Autovector {i + 1}:", autovectores[:, i])

Autovalores: [2.5+0.8660254j 2.5-0.8660254j]
Autovectores:
Autovector 1: [0.35355339-0.61237244j 0.70710678+0.j        ]
Autovector 2: [0.35355339+0.61237244j 0.70710678-0.j        ]


In [4]:
# Comprobar que son efectivamente autovalores y autovectores
for i in range(len(autovalores)):
    lambda_i = autovalores[i]
    v_i = autovectores[:, i]
    result = np.dot(T_matrix, v_i)
    if np.allclose(result, lambda_i * v_i):
        print(f"El autovector {i + 1} es un autovector con autovalor {lambda_i}.")
    else:
        print(f"El autovector {i + 1} NO es un autovector con autovalor {lambda_i}.")


El autovector 1 es un autovector con autovalor (2.5+0.8660254037844386j).
El autovector 2 es un autovector con autovalor (2.5-0.8660254037844386j).


In [5]:
T_matrix = np.array([[2, 1],
                     [1, 1]])

autovalores, autovectores = np.linalg.eig(T_matrix)

print("Autovalores:", autovalores)
print("Autovectores:")
for i in range(len(autovectores)):
    print(f"Autovector {i + 1}:", autovectores[:, i])

Autovalores: [2.61803399 0.38196601]
Autovectores:
Autovector 1: [0.85065081 0.52573111]
Autovector 2: [-0.52573111  0.85065081]


In [6]:
for i in range(len(autovalores)):
    lambda_i = autovalores[i]
    v_i = autovectores[:, i]
    result = np.dot(T_matrix, v_i)
    if np.allclose(result, lambda_i * v_i):
        print(f"El autovector {i + 1} es un autovector con autovalor {lambda_i}.")
    else:
        print(f"El autovector {i + 1} NO es un autovector con autovalor {lambda_i}.")

El autovector 1 es un autovector con autovalor 2.618033988749895.
El autovector 2 es un autovector con autovalor 0.3819660112501052.
